pkg/resource/deploy/source_eval.go

```diff
@@ -1605,13 +1644,42 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 		parsedAliases = append(parsedAliases, alias)
 	}
 
-	dependencies := []resource.URN{}
-	for _, dependingURN := range opts.GetDependsOn() {
-		urn, err := resource.ParseURN(dependingURN)
-		if err != nil {
-			return nil, rpcerror.New(codes.InvalidArgument, fmt.Sprintf("invalid dependency URN: %s", err))
+	// Reparse the dependency information from any transformation results
+	if len(req.Transforms) > 0 {
+		dependencies = mapset.NewSet[resource.URN]()
+		for _, dependingURN := range opts.DependsOn {
+			urn, err := resource.ParseURN(dependingURN)
+			if err != nil {
+				return nil, rpcerror.New(codes.InvalidArgument, fmt.Sprintf("invalid dependency URN: %s", err))
+			}
+			dependencies.Add(urn)
+		}
+		// Now we've run the transforms we can rebuild the property dependency maps. If we have output values we can add the
+		// dependencies from them to the dependencies map we send to the provider and save to state.
+		propertyDependencies = make(map[resource.PropertyKey]mapset.Set[resource.URN])
+		for key, output := range props {
+			deps := mapset.NewSet[resource.URN]()
+			addOutputDependencies(deps, output)
+			propertyDependencies[key] = deps
+
+			// Also add these to the overall dependencies
+			dependencies = dependencies.Union(deps)
+		}
+	} else {
+		// If we ran transforms we would have merged all the dependencies togther already, but if we didn't we want to
+		// ensure any output values add their dependencies to the dependencies map we send to the provider.
+		for key, output := range props {
```

Comment:
I'm not sure I'm following this change properly.  Since we didn't upgrade the props in the case where there are no transforms, couldn't we just use the old code from https://github.com/pulumi/pulumi/pull/15637/files#diff-984b3279d61b588030474831898bb76d093d863ca2d0015bb081d842602a9ab7L1665?  Or why is this change necessary?

Reply:
This is pretty much that code but for Set instead of slice.